fe/fe-core/src/main/java/org/apache/doris/fs/obj/S3ObjStorage.java

```
@@ -249,9 +300,26 @@ public Status copyObject(String origFilePath, String destFilePath) {
     public RemoteObjects listObjects(String absolutePath, String continuationToken) throws DdlException {
         try {
             S3URI uri = S3URI.create(absolutePath, forceHostedStyle);
+            String bucket = uri.getBucket();
             String prefix = uri.getKey();
-            ListObjectsV2Request.Builder requestBuilder = ListObjectsV2Request.builder().bucket(uri.getBucket())
-                        .prefix(normalizePrefix(prefix));
+            if (!StringUtils.isEmpty(uri.getVirtualBucket())) {
+                // Support s3 compatible service. The generated HTTP request for list objects likes:
+                //
+                //  GET /<bucket-name>?list-type=2&prefix=<prefix>
+                prefix = bucket + "/" + prefix;
+                String endpoint = properties.get(S3Properties.ENDPOINT);
+                if (endpoint.contains("cos.")) {
+                    bucket = "/";
+                } else if (endpoint.contains("oss-")) {
+                    bucket = uri.getVirtualBucket();
+                } else if (endpoint.contains("obs.")) {
```

maybe throw an exception